### PR TITLE
Adding support for predict and predict_proba in mlflow recipe model

### PR DESCRIPTION
**Pull request by sunishsheth2009**
## What changes are proposed in this pull request?
Adding support for `predict` and `predict_proba` in mlflow recipe model

## How is this patch tested?
- [x] Test passes
- [x] Checked in notebook

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Adding support for `predict` and `predict_proba` in mlflow recipe model

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/recipes/utils/wrapped_recipe_model.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+from mlflow.pyfunc import PythonModel
+import pandas as pd
+import numpy as np
+
+
+class WrappedRecipeModel(PythonModel):
+    def __init__(self, classifier, predict_scores_for_all_classes=True, prefix="predicted_"):
+        super().__init__()
+        self._classifier = classifier
```
Comment (dbczumar) [inline mlflow/recipes/utils/wrapped_recipe_model.py:9]: The `classifier` shouldn't be part of the constructor, since this will pickle the classifier along with the model class (this may work okay for scikit-learn but doesn't work for many ML model types that can't be pickled)

Instead, we should pass in a *reference* to the classifier at model save time via the `artifacts` parameter to `mlflow.pyfunc.log_model()` and then load it via the `PythonModel.load_context()` method: Here's an example -  https://mlflow.org/docs/latest/models.html#example-saving-an-xgboost-model-in-mlflow-format

File: mlflow/recipes/utils/wrapped_recipe_model.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        self.predict_scores_for_all_classes = predict_scores_for_all_classes
+        self.prefix = prefix
+        self.classification = False
+
+    def predict(self, model_input):
```
Comment (dbczumar) [inline mlflow/recipes/utils/wrapped_recipe_model.py:14]: `predict()` takes three arguments: `self, context, model_input`

File: mlflow/recipes/steps/train.py [hunk 3]
```diff
@@ -162,7 +167,10 @@ def _validate_and_apply_step_config(self):
                 error_code=INVALID_PARAMETER_VALUE,
             )
 
-        self.predict_proba = self.step_config.get("predict_proba", False)
+        self.predict_prefix = self.step_config.get("predict_prefix", "predicted_")
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:170]: Just to confirm, have other stakeholders approved this prefix conf for the model `predict()` output names? Just making sure since it's not obvious to me that this needs to be configurable.

Comment (dbczumar) [reply mlflow/recipes/steps/train.py:170]: cc @jinzhang21

File: mlflow/recipes/steps/train.py [hunk 4]
```diff
@@ -329,40 +337,69 @@ def my_warn(*args, **kwargs):
...
-
-                if self.predict_proba:
-                    mlflow.sklearn.save_model(
-                        model, output_model_path, pyfunc_predict_fn="predict_proba"
+                sklearn_model_uri = get_step_output_path(
+                    recipe_root_path=self.recipe_root,
+                    step_name=self.name,
+                    relative_path=TrainStep.SKLEARN_MODEL_ARTIFACT_RELATIVE_PATH,
+                )
+                if os.path.exists(model_uri) and os.path.isdir(model_uri):
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:358]: Shouldn't we still remove it even if it's not a directory?

File: mlflow/recipes/steps/train.py [hunk 5]
```diff
@@ -329,40 +337,69 @@ def my_warn(*args, **kwargs):
...
+                    shutil.rmtree(model_uri)
+                if os.path.exists(sklearn_model_uri) and os.path.isdir(sklearn_model_uri):
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:360]: Ditto here (https://github.com/mlflow/mlflow/pull/7474/files#r1052714099)

File: mlflow/recipes/steps/train.py [hunk 6]
```diff
@@ -329,40 +337,69 @@ def my_warn(*args, **kwargs):
...
+                    shutil.rmtree(sklearn_model_uri)
+
+                # Saving the sklearn model along with its pyfunc model since mlflow evaluate
+                # predicts results using an sklearn model
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:364]:
```suggestion
                # Saving the sklearn model as a separate output since `mlflow.evaluate()`, which is
                # used in the evaluate step of the recipe, needs this model's pyfunc representation to
                # computes metrics (the pyfunc representation of the user-facing model logged to
                # MLflow Tracking is not currently compatible with `mlflow.evaluate()`)
```

File: mlflow/recipes/steps/train.py [hunk 7]
```diff
@@ -327,40 +335,69 @@ def my_warn(*args, **kwargs):
...
-                        model,
-                        f"{self.name}/model",
-                        signature=model_schema,
-                        code_paths=self.code_paths,
-                    )
-                output_model_path = get_step_output_path(
+
+                trained_pipeline = make_pipeline(transformer, estimator)
+                # Creating a wrapped recipe model for the pipeline so users can achieve both
+                # predict and predict_proba(for a classification problem) correctly.
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:341]: The wrapped model still only exposes one API (`predict()`), right? Can we rephrase this to indicate that the wrapped model is created so that `predict()` can output class predictions and probabilities at the same time?

File: mlflow/recipes/steps/train.py [hunk 8]
```diff
@@ -327,40 +335,69 @@ def my_warn(*args, **kwargs):
...
+                        signature=model_schema,
+                        code_path=self.code_paths,
+                    )
+                model = mlflow.pyfunc.load_model(model_uri)
+                model_info = Model.load(model_uri)
+                model_data_subpath = os.path.join(
+                    "artifacts", TrainStep.SKLEARN_MODEL_ARTIFACT_RELATIVE_PATH, "model.pkl"
+                )
+                # Adding a sklearn flavor to the pyfunc model so models could be loaded easily
+                # using mlflow.sklearn.load_model
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:391]: Can we move this comment up above line 385? `model_info = Model.load(model_uri)`

File: mlflow/recipes/steps/train.py [hunk 9]
```diff
@@ -327,40 +335,69 @@ def my_warn(*args, **kwargs):
...
+                if os.path.exists(sklearn_model_uri) and os.path.isdir(sklearn_model_uri):
+                    shutil.rmtree(sklearn_model_uri)
+
+                # Saving the sklearn model along with its pyfunc model since mlflow evaluate
+                # predicts results using an sklearn model
+                mlflow.sklearn.save_model(trained_pipeline, sklearn_model_uri)
+                artifacts = {"model_path": sklearn_model_uri}
+                with TempDir() as tmp:
+                    # Saving a temp model so that signature of the model output could be inferred
+                    # before the pyfunc model can be saved as the specified model_uri location
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:367]:
```suggestion
                    # Saving a temp model so that the output schema (signature) of the model's pyfunc
                    # representation can be inferred and included when logging the model to MLflow Tracking.
                    # Unfortunately, there is currently no easy way to infer the model's signature without first saving
                    # a copy of it, and there is no easy way to add an inferred signature to an existing model
```

File: mlflow/recipes/steps/train.py [hunk 10]
```diff
@@ -160,7 +165,10 @@ def _validate_and_apply_step_config(self):
                 error_code=INVALID_PARAMETER_VALUE,
             )
 
-        self.predict_proba = self.step_config.get("predict_proba", False)
+        self.predict_prefix = self.step_config.get("predict_prefix", "predicted_")
+        self.predict_scores_for_all_classes = self.step_config.get(
+            "predict_scores_for_all_classes", True
+        )
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:171]: Same with this conf - have stakeholders signed off? cc @jinzhang21

File: mlflow/recipes/utils/wrapped_recipe_model.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        super().__init__()
+        self.predict_scores_for_all_classes = predict_scores_for_all_classes
+        self.prefix = prefix
+
+    def load_context(self, context):
+        self._classifier = mlflow.sklearn.load_model(context.artifacts["model_path"])
+
+    def predict(self, context, model_input):
+        predicted_label = self._classifier.predict(model_input)
+        if not hasattr(self._classifier, "classes_"):
```
Comment (dbczumar) [inline mlflow/recipes/utils/wrapped_recipe_model.py:18]: Can we leave an inline comment here explaining why we check for this attribute?

File: mlflow/recipes/steps/train.py [hunk 12]
```diff
@@ -398,16 +435,22 @@ def my_warn(*args, **kwargs):
 
             target_data = raw_validation_df[self.target_col]
             prediction_result = model.predict(raw_validation_df.drop(self.target_col, axis=1))
-            if self.predict_proba:
-                prediction_result_probs = model.predict_proba(
-                    raw_validation_df.drop(self.target_col, axis=1)
+
+            use_predict_proba = isinstance(prediction_result, pd.DataFrame)
+            if use_predict_proba:
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:440]: Isn't it possible for the wrapped model to output a DataFrame even if the model doesn't have a `classes_` attribute? Is there a more specific check we can use?

File: mlflow/recipes/utils/wrapped_recipe_model.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+        try:
+            probabilities = self._classifier.predict_proba(model_input)
+            if self.predict_scores_for_all_classes:
+                output = pd.DataFrame(columns=score_cols, data=probabilities)
+            output[f"{self.prefix}score"] = np.max(probabilities, axis=1)
+            output[f"{self.prefix}label"] = predicted_label
+
+            return output
+        except Exception:
+            # swallow that some models don't have predict_proba.
```
Comment (dbczumar) [inline mlflow/recipes/utils/wrapped_recipe_model.py:33]: Shouldn't we check to see if `hasattr(self._classifier, "predict_proba")`? That sounds better than try / catching all Exceptions

File: tests/recipes/test_train_step.py [hunk 14]
```diff
@@ -677,3 +694,62 @@ def _import_module(name: str, package: str = None):
...
+        recipe_config = read_yaml(tmp_recipe_root_path, _RECIPE_CONFIG_FILE_NAME)
+        train_step = TrainStep.from_recipe_config(recipe_config, str(tmp_recipe_root_path))
+        train_step.run(str(train_step_output_dir))
+
+    model_uri = get_step_output_path(
+        recipe_root_path=tmp_recipe_root_path,
+        step_name="train",
+        relative_path=TrainStep.MODEL_ARTIFACT_RELATIVE_PATH,
+    )
+    model = mlflow.pyfunc.load_model(model_uri)
```
Comment (dbczumar) [inline tests/recipes/test_train_step.py:741]: Looks like we don't have any test coverage verifying that the model can be loaded with `mlflow.sklearn.load_model()`. Can we add some?

File: mlflow/recipes/steps/train.py (hunk 3 above)
Comment (sunishsheth2009) [reply mlflow/recipes/steps/train.py:168]: There is a P1 requirement:
```
[P1] I can set a prefix of the output column names to avoid name collision with input columns.
Example: prefix=”my_”, output becomes {“my_label”: “true”, “my_score”: 0.85}
“predicted_” is the default prefix.
```
So added this conf based on that. 

We don't need to document them right away since we have sensible defaults for it. :)

File: tests/recipes/test_train_step.py [hunk 15]
```diff
@@ -677,3 +694,62 @@ def _import_module(name: str, package: str = None):
...
+    )
+
+    output = model.predict(validation_dataset)
+
+    assert list(output.columns) == [
+        "predicted_score_a",
+        "predicted_score_b",
+        "predicted_score",
+        "predicted_label",
+    ]
```
Comment (dbczumar) [inline tests/recipes/test_train_step.py:755]: Can we assert that the values are reasonable as well?

File: mlflow/recipes/steps/train.py (hunk 10 above)
Comment (sunishsheth2009) [reply mlflow/recipes/steps/train.py:171]: There is a P1 requirement:
```
[P1] I can configure it to always predict the scores for both labels. So I can easily get the score of a certain label, e.g., the probability of click (positive) in CTR prediction.
Example: {“predicted_label”: “true”, “predicted_score”: 0.85, “predicted_score_true”: 0.85, “predicted_score_false”: 0.15}
```
So added this conf based on that. 

We don't need to document them right away since we have sensible defaults for it. :)

File: mlflow/recipes/steps/train.py (hunk 12 above)
Comment (sunishsheth2009) [reply mlflow/recipes/steps/train.py:446]: Currently the wrapped model would only output a DF if it has `classes_` or else it would output the label directly rather than a df with probabilities, label and score. 
I couldn't find a better alternative for the same. Have any suggestions?

File: mlflow/recipes/steps/train.py (hunk 3 above)
Comment (dbczumar) [reply mlflow/recipes/steps/train.py:168]: Sounds good!

File: mlflow/recipes/steps/train.py (hunk 10 above)
Comment (dbczumar) [reply mlflow/recipes/steps/train.py:171]: Sounds good!

File: mlflow/recipes/steps/train.py [hunk 16]
```diff
@@ -398,16 +441,22 @@ def my_warn(*args, **kwargs):
...
+                prediction_result_for_error = (
+                    prediction_result.drop(
+                        [f"{self.predict_prefix}label", f"{self.predict_prefix}score"], axis=1
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:449]: Can we check to make sure that these columns exist before attempting to drop / access them?

File: mlflow/recipes/steps/train.py (hunk 12 above)
Comment (dbczumar) [reply mlflow/recipes/steps/train.py:446]: What if I pass in a dataframe with multiple rows as an input to self._classifier.predict()? The output wouldn’t be a scalar, right? Would it be a list? Numpy array? Are we certain that it’s not a DataFrame?

File: mlflow/recipes/steps/train.py [hunk 17]
```diff
@@ -418,18 +467,22 @@ def my_warn(*args, **kwargs):
                 }
             )
             train_predictions = model.predict(raw_train_df.drop(self.target_col, axis=1))
-            if self.predict_proba:
-                train_predicted_probs = model.predict_proba(
-                    raw_train_df.drop(self.target_col, axis=1)
+            if isinstance(train_predictions, pd.DataFrame):
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:470]: Ditto here: https://github.com/mlflow/mlflow/pull/7474/files#r1052721048

File: mlflow/recipes/steps/train.py [hunk 18]
```diff
@@ -418,18 +467,22 @@ def my_warn(*args, **kwargs):
...
+                train_predicted_result = model.predict(raw_train_df.drop(self.target_col, axis=1))
+                train_predicted_probs = train_predicted_result.drop(
+                    [f"{self.predict_prefix}label", f"{self.predict_prefix}score"], axis=1
                 )
                 predicted_training_data = raw_train_df.assign(
-                    predicted_data=train_predictions,
-                    predicted_probability=train_predicted_probs[:, 0],
+                    predicted_data=train_predictions[f"{self.predict_prefix}label"],
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:476]: Ditto here: https://github.com/mlflow/mlflow/pull/7474/files#r1052799413

File: mlflow/recipes/steps/train.py [hunk 19]
```diff
@@ -418,18 +470,25 @@ def my_warn(*args, **kwargs):
...
+                    predicted_probability=train_predicted_probs.iloc[0:, 0].values,
                 )
                 worst_examples_df = BaseStep._generate_worst_examples_dataframe(
                     raw_train_df,
-                    train_predicted_probs[:, 0],
-                    error_fn(train_predicted_probs, raw_train_df[self.target_col].to_numpy()),
+                    train_predicted_probs.iloc[0:, 0].values,
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:487]: Small question: What's the difference between `[:, 0]` (previous behavior) and `[0:, 0]`?

File: mlflow/recipes/steps/train.py [hunk 20]
```diff
@@ -418,18 +470,25 @@ def my_warn(*args, **kwargs):
...
+                    error_fn(
+                        train_predicted_probs.iloc[0:].values,
```
Comment (dbczumar) [inline mlflow/recipes/steps/train.py:489]: Would this work instead?

```suggestion
                        train_predicted_probs.values,
```

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @sunishsheth2009 !

File: mlflow/recipes/steps/train.py (hunk 19 above)
Comment (sunishsheth2009) [reply mlflow/recipes/steps/train.py:487]: Its the same, I just wanted to make it consistent with the code below.

File: mlflow/recipes/utils/wrapped_recipe_model.py [hunk 21]
```diff
@@ -0,0 +1,35 @@
+import mlflow
+from mlflow.pyfunc import PythonModel
+import pandas as pd
+import numpy as np
+
+
+class WrappedRecipeModel(PythonModel):
```
Comment (jinzhang21) [inline mlflow/recipes/utils/wrapped_recipe_model.py:7]: Is this model only meant for classification? Its name suggests it's universal. Can we rename it to `RecipeClassificationModel` or something?

File: mlflow/recipes/utils/wrapped_recipe_model.py [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+    def __init__(self, predict_scores_for_all_classes=True, prefix="predicted_"):
+        super().__init__()
+        self.predict_scores_for_all_classes = predict_scores_for_all_classes
+        self.prefix = prefix
```
Comment (jinzhang21) [inline mlflow/recipes/utils/wrapped_recipe_model.py:11]: Is this `prefix` the same of `predict_prefix` in the step config? Prefer to have the same name to avoid confusion.

File: mlflow/recipes/utils/wrapped_recipe_model.py [hunk 23]
```diff
@@ -0,0 +1,35 @@
+import mlflow
+from mlflow.pyfunc import PythonModel
+import pandas as pd
+import numpy as np
+
+
+class WrappedRecipeModel(PythonModel):
+    def __init__(self, predict_scores_for_all_classes=True, prefix="predicted_"):
```
Comment (jinzhang21) [inline mlflow/recipes/utils/wrapped_recipe_model.py:8]: Do `predict_scores_for_all_classes` and `prefix` need to have default value? It should be set by the caller to make sure the value is passed down correctly.

File: mlflow/recipes/utils/wrapped_recipe_model.py (hunk 21 above)
Comment (jinzhang21) [reply mlflow/recipes/utils/wrapped_recipe_model.py:7]: Shouldn't a base class + subclass approach work better? Almost all contexts in this class suggest it's meant for a classifier model. I don't see it can be easily extended for "more options" in the future.

**Review by jinzhang21 — APPROVED** [review]

File: mlflow/recipes/utils/wrapped_recipe_model.py (hunk 21 above)
Comment (sunishsheth2009) [reply mlflow/recipes/utils/wrapped_recipe_model.py:7]: I think once we have a broader usecase, we can design for it better. Hopefully we won't need all of this once we have microsoft proposal merged as well to add additional functions to pyfunc model
